### ISSUE TITLE
Cannot customize username of OIDC authentication

### ISSUE THREAD

**Issue by kazuki43zoo** [spring-projects/spring-security#4861]
### Summary

I want to use `name` claim instead of `sub` claim as principal username (e.g. In google, `sub` claim value is 21 length numeric). But current implementation, it does not allow in the `OidcUserService`.

### Actual Behavior

Use `sub` claim value (= cannot apply customization).

### Expected Behavior

Use claim value of the `UserInfoEndpoint#userNameAttributeName`.

### Configuration

```java
package com.example.demooauth2login;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
import org.springframework.security.config.oauth2.client.CommonOAuth2Provider;
import org.springframework.security.oauth2.client.InMemoryOAuth2AuthorizedClientService;
import org.springframework.security.oauth2.client.OAuth2AuthorizedClientService;
import org.springframework.security.oauth2.client.registration.ClientRegistrationRepository;
import org.springframework.security.oauth2.client.registration.InMemoryClientRegistrationRepository;

@Configuration
public class SecurityConfig extends WebSecurityConfigurerAdapter {

	@Override
	protected void configure(HttpSecurity http) throws Exception {
		http.authorizeRequests()
				.anyRequest().authenticated();
		http.oauth2Login()
				.clientRegistrationRepository(clientRegistrationRepository())
				.authorizedClientService(oAuth2AuthorizedClientService());
	}

	@Bean
	public ClientRegistrationRepository clientRegistrationRepository() {
		return new InMemoryClientRegistrationRepository(
				CommonOAuth2Provider.GOOGLE.getBuilder("google")
					.clientId("your client id")
					.clientSecret("your client secret")
					.userNameAttributeName("name")  // ### Chane to 'name' from 'sub'
					.build()
		);
	}

	@Bean
	public OAuth2AuthorizedClientService oAuth2AuthorizedClientService() {
		return new InMemoryOAuth2AuthorizedClientService(clientRegistrationRepository());
	}

}
```

Note:

In Spring Boot 2.0.0.M6, configuration property does not apply to a `Builder` of `CleintRegistration` (See spring-projects/spring-boot#11149 ).

```properties
spring.security.oauth2.client.registration.google.clientId=your client id
spring.security.oauth2.client.registration.google.clientSecret=your client secret
spring.security.oauth2.client.provider.google.userNameAttribute=name  # This configuration does not apply into Spring Security's component
```

### Version

* Spring Security 5.0.0.RC1
* Spring Boot 2.0.0.M6

**Comment by kazuki43zoo** [comment]
I will suggest fix on `OidcUserService` as follow:

```diff
@Override
public OidcUser loadUser(OidcUserRequest userRequest) throws OAuth2AuthenticationException {
	// ...
+	String userNameAttributeName = userRequest.getClientRegistration().getProviderDetails().getUserInfoEndpoint().getUserNameAttributeName();
+	if (StringUtils.hasText(userNameAttributeName)) {
+		return new DefaultOidcUser(authorities, userRequest.getIdToken(), userInfo, userNameAttributeName);
+	} else {
		return new DefaultOidcUser(authorities, userRequest.getIdToken(), userInfo);
+	}
}
```

WDYT?

**Comment by kazuki43zoo** [comment]
Already fix via #4812.
Thanks!
